src/sagas/task.saga.js

```diff
@@ -0,0 +1,339 @@
+
+import type { Task, ProjectType } from '../types';
+import type { Saga } from 'redux-saga';
+
+const chalk = new chalkRaw.constructor({ level: 3 });
+
+export function* launchDevServer({ task }: { task: Task }): Saga<void> {
+  const project = yield select(getProjectById, task.projectId);
+  const projectPath = yield select(getPathForProjectId, task.projectId);
+
+  try {
+    const port = yield call(findAvailablePort);
+    const { args, env } = yield call(
+      getDevServerCommand,
+      task,
+      project.type,
+      port
+    );
+
+    const child = yield call(
+      [childProcess, childProcess.spawn],
+      PACKAGE_MANAGER_CMD,
+      args,
+      {
+        cwd: projectPath,
+        env: { ...getBaseProjectEnvironment(projectPath), ...env },
+      }
+    );
+
+    // Now that we have a port/processId for the server, attach it to
+    // the task. The port is used for opening the app, the pid is used
+    // to kill the process
+    yield put(attachTaskMetadata(task, child.pid, port));
+
+    yield call([ipcRenderer, ipcRenderer.send], 'addProcessId', child.pid);
+
+    const stdioChannel = createStdioChannel(child, {
+      stdout: emitter => data => {
+        // Ok so, unfortunately, failure-to-compile is still pushed
+        // through stdout, not stderr. We want that message specifically
+        // to trigger an error state, and so we need to parse it.
+        const text = data.toString();
+
+        const isError = text.includes('Failed to compile');
+
+        emitter({ channel: 'stdout', text, isError });
+      },
+      stderr: emitter => data => {
+        emitter({ channel: 'stderr', text: data.toString() });
+      },
+      exit: emitter => code => {
+        // For Windows Support
+        // Windows sends code 1 (I guess its because we force kill??)
+        const successfulCode = isWin ? 1 : 0;
+        const wasSuccessful = code === successfulCode || code === null;
+        const timestamp = new Date();
+
+        emitter({ channel: 'exit', timestamp, wasSuccessful });
+        // calling emitter(END) will break out of the try block of any
+        // actively listening subscribers when they take() it
+        emitter(END);
+      },
+    });
+
+    try {
+      while (true) {
+        const message = yield take(stdioChannel);
+
+        // eslint-disable-next-line default-case
+        switch (message.channel) {
+          case 'stdout':
+            yield put(
+              receiveDataFromTaskExecution(task, message.text, message.isError)
+            );
+            break;
+          case 'stderr':
+            yield put(receiveDataFromTaskExecution(task, message.text));
+            break;
+          case 'exit':
+            yield call(displayTaskComplete, task);
+            yield put(
+              completeTask(task, message.timestamp, message.wasSuccessful)
+            );
+            break;
+        }
+      }
+    } finally {
+      /* child process has completed */
+    }
```

Comment:
I think we can omit the `finally` block, unless there are short-term plans to use it?